server/channels/app/notification_test.go

```diff
@@ -212,6 +216,268 @@ func TestSendNotifications(t *testing.T) {
+		require.Equal(t, model.WebsocketEventPosted, received1.EventType())
+		assert.Equal(t, `["`+th.BasicUser.Id+`"]`, received1.GetData()["mentions"])
+
+		received2 = <-messages2
+		require.Equal(t, model.WebsocketEventPosted, received2.EventType())
+		assert.Nil(t, received2.GetData()["mentions"])
+	})
+
+	t.Run("should inform each user in a group if they were mentioned by a post", func(t *testing.T) {
+		th.App.Srv().SetLicense(getLicWithSkuShortName(model.LicenseShortSkuEnterprise))
+
+		// Make a group and add users
+		group := th.CreateGroup()
+		group.AllowReference = true
+		group, updateErr := th.App.UpdateGroup(group)
+		require.Nil(t, updateErr)
+
+		_, upsertErr := th.App.UpsertGroupMember(group.Id, th.BasicUser.Id)
+		require.Nil(t, upsertErr)
+		_, upsertErr = th.App.UpsertGroupMember(group.Id, th.BasicUser2.Id)
+		require.Nil(t, upsertErr)
+
+		// Set up the websockets
+		messages1, closeWS1 := connectFakeWebSocket(t, th, th.BasicUser.Id, "")
+		defer closeWS1()
+
+		messages2, closeWS2 := connectFakeWebSocket(t, th, th.BasicUser2.Id, "")
+		defer closeWS2()
+
+		// Confirm permissions for group mentions are correct
+		post := &model.Post{
+			UserId:    sender.Id,
+			ChannelId: th.BasicChannel.Id,
+			Message:   "@" + *group.Name,
+		}
+		require.True(t, th.App.allowGroupMentions(th.Context, post))
+
+		// Test sending notifications
+		_, err := th.App.SendNotifications(th.Context, post, th.BasicTeam, th.BasicChannel, th.BasicUser, nil, false)
+		require.NoError(t, err)
+
+		received1 := <-messages1
+		require.Equal(t, model.WebsocketEventPosted, received1.EventType())
+		assert.Equal(t, `["`+th.BasicUser.Id+`"]`, received1.GetData()["mentions"])
+
+		received2 := <-messages2
+		require.Equal(t, model.WebsocketEventPosted, received2.EventType())
+		assert.Equal(t, `["`+th.BasicUser2.Id+`"]`, received2.GetData()["mentions"])
+	})
+
+	t.Run("should inform each user if they are following a thread that was posted in", func(t *testing.T) {
+		messages1, closeWS1 := connectFakeWebSocket(t, th, th.BasicUser.Id, "")
+		defer closeWS1()
+
+		messages2, closeWS2 := connectFakeWebSocket(t, th, th.BasicUser2.Id, "")
+		defer closeWS2()
+
+		// First post mentioning the whole channel
+		post := &model.Post{
+			UserId:    sender.Id,
+			ChannelId: th.BasicChannel.Id,
+			Message:   "@channel",
+		}
+		_, err := th.App.SendNotifications(th.Context, post, th.BasicTeam, th.BasicChannel, th.BasicUser, nil, false)
+		require.NoError(t, err)
+
+		received1 := <-messages1
+		require.Equal(t, model.WebsocketEventPosted, received1.EventType())
+		assert.Equal(t, `["`+th.BasicUser.Id+`"]`, received1.GetData()["mentions"])
+
+		received2 := <-messages2
+		require.Equal(t, model.WebsocketEventPosted, received2.EventType())
+		assert.Equal(t, `["`+th.BasicUser2.Id+`"]`, received2.GetData()["mentions"])
+	})
+}
+
+func connectFakeWebSocket(t *testing.T, th *TestHelper, userID string, connectionID string) (chan *model.WebSocketEvent, func()) {
```

Comment:
Any reason we cannot copy over `th.CreateWebSocketClient` from api layer to this? See `TestPostReminder` for example. This code is re-doing lot of the initialization code that happens inside the web_conn.go file. Any changes to that, will need to be copied over to this again. I'd rather have something high level to interact with the websockets.

Reply:
I can see if I can copy that over. I don't know if we can access the API endpoint here, but it'd be nice if we can use the same interface as that method at least

Reply:
I spent a couple hours trying to make this work, and I couldn't manage to make this code use a `model.WebSocketClient` since it makes a lot of assumptions about there being an HTTP API to connect the WebSocket. This function also replicates the logic from `api4/websocket.go` which sets connects the `WebConn` and isn't part of the `app` package, so we'd either need to move that to the app layer or duplicate it here. I've spent enough time on this that I think I need to move on from this one